storage/innobase/include/fts0types.inl

```diff
@@ -75,22 +75,28 @@ fts_ranking_doc_id_cmp(
 	const fts_ranking_t*	rk1 = (const fts_ranking_t*) p1;
 	const fts_ranking_t*	rk2 = (const fts_ranking_t*) p2;
 
-	return((int)(rk1->doc_id - rk2->doc_id));
+	return(fts_doc_id_cmp(&rk1->doc_id, &rk2->doc_id));
 }
 
 /******************************************************************//**
 Compare two doc_ids.
 @return < 0 if n1 < n2, 0 if n1 == n2, > 0 if n1 > n2 */
 UNIV_INLINE
 int fts_doc_id_cmp(
-/*==================*/
+/*===============*/
 	const void*	p1,			/*!< in: id1 */
 	const void*	p2)			/*!< in: id2 */
 {
 	const doc_id_t*	up1 = static_cast<const doc_id_t*>(p1);
 	const doc_id_t*	up2 = static_cast<const doc_id_t*>(p2);
 
-	return static_cast<int>(*up1 - *up2);
+	if (*up1 < *up2) {
+		return (-1);
+	} else if (*up1 > *up2) {
+		return (1);
+	} else {
+		return (0);
+	}
```

Comment:
I was curious about the generated code, so I tested the following on https://godbolt.org:
```c++
int cmp(int a) { return a < 0 ? -1 : a > 0; }
int cmp2(unsigned long long a, unsigned long long b) { if (a < b) return -1; if (a > b) return 1; return 0; }
```
Up to GCC 7, for the unary function `cmp` the compiler would generate conditional branches on IA-32, AMD64, ARMv7 and ARMv8. Starting with GCC 8, it would generate some conditional instructions.

For the binary function `cmp2` the code is rather acceptable already on GCC 4.8.5 for AMD64, which is the oldest that we currently support:
```asm
        xor     edx, edx
        cmp     rdi, rsi
        mov     eax, -1
        seta    dl
        cmovnb  eax, edx
        ret
```
On ARMv7 as well as ARMv8, no conditional branches were emitted. On POWER, I guess that there are no conditional ALU instructions, because whichever compiler version I tested, there would be conditional branches.

Reply:
I checked this a little deeper. There are quite a few references to `ib_rbt_node_t`, and there is a `fts_cache_t::total_size` that will have to be maintained. It might not be easy to replace this data structure with the C++11 `std::multimap` while retaining that bookkeeping.

What we can easily do is to remove `ib_vector_sort` and `fts_doc_id_cmp` (30dbb1ad36cce9752d597484b68bf91ac61ef1c3). The generated code for my replacement `fts_doc_ids_sort` as well as its callees looked reasonable to me.